components/browser/engine-gecko-nightly/src/main/java/mozilla/components/browser/engine/gecko/NestedGeckoView.kt

```diff
@@ -87,23 +83,46 @@ open class NestedGeckoView(context: Context) : GeckoView(context), NestedScrolli
             }
 
             MotionEvent.ACTION_DOWN -> {
+                // A new gesture started. Reset handled status and ask GV if it can handle this.
+                inputResult = INPUT_RESULT_UNHANDLED
+                updateInputResult(event)
+
+                nestedOffsetY = 0
                 lastY = eventY
-                startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL)
+
+                // The event should be handled either by onTouchEvent,
+                // either by onTouchEventForResult, never by both.
+                // Early return if we sent it to updateInputResult(..) which calls onTouchEventForResult.
+                event.recycle()
+                return true
             }
 
             // We don't care about other touch events
             MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> stopNestedScroll()
         }
 
+        // Execute event handler from parent class in all cases
+        val eventHandled = callSuperOnTouchEvent(event)
+
         // Recycle previously obtained event
         event.recycle()
 
-        return inputResult
+        return eventHandled
+    }
+
+    @VisibleForTesting
+    internal fun callSuperOnTouchEvent(event: MotionEvent): Boolean {
+        return super.onTouchEvent(event)
     }
 
-    // Helper function to make testing of this method easier
-    internal fun handleEvent(event: MotionEvent): Int {
-        return super.onTouchEventForResult(event)
+    @VisibleForTesting
+    internal fun updateInputResult(event: MotionEvent) {
+        rootView.toScope().launch {
+            super.onTouchEventForResult(event).await()?.let {
+                inputResult = it
+                startNestedScroll(ViewCompat.SCROLL_AXIS_VERTICAL)
+            }
+        }
```

Comment:
It looks like this gets called from `onTouchEvent()` and therefore potentially quite often? Under the hood `toScope()` always creates a new `MainScope` and registers an `OnAttachStateChangeListener` to know when to cancel it. So we may end up creating a lot of them. Maybe we can create a scope from the View only once inside this class?